### PR TITLE
Small cleanup to FieldValue sentinel code to make it easier to add more sentinels in the future.

### PR DESCRIPTION
**Pull request by mikelehen**
This technically includes a small breaking change. Before we would allow you to do:

`doc.set({a: [0, FieldValue.delete(), 2] }, {merge: true})`

Which ended up equivalent to:

`doc.set({a: [0, null, 2] }, {merge: true})`

But with my change it will now result in an error:

`FieldValue.delete() is not currently supported inside arrays.`

Let me know if you think that's an issue. I can rework it but I don't think the existing behavior was particularly useful or intentional.

### CODE REVIEW

File: packages/firestore/src/api/user_data_converter.ts [hunk 1]
```diff
@@ -530,51 +530,60 @@ export class UserDataConverter {
     } else if (value instanceof DocumentKeyReference) {
       return new RefValue(value.databaseId, value.key);
     } else if (value instanceof FieldValueImpl) {
```
Comment (schmidt-sebastian) [inline packages/firestore/src/api/user_data_converter.ts]: Just a suggestion - we could move this block up to `parseData`. Then we would have a single place in the code that decides between our parseX implementations.

File: packages/firestore/src/api/user_data_converter.ts [hunk 2]
```diff
@@ -530,51 +530,60 @@ export class UserDataConverter {
...
+  }
+
+  private parseSentinelFieldValue(
+    value: FieldValueImpl,
+    context: ParseContext
+  ): FieldValue | null {
+    // Sentinels are only supported with writes, and not within arrays.
+    if (!isWrite(context.dataSource)) {
+      throw context.createError(
+        `${value.methodName} can only be used with update() and set().`
```
Comment (schmidt-sebastian) [inline packages/firestore/src/api/user_data_converter.ts]: Nit: Remove period to match other errors.

File: packages/firestore/test/integration/api/validation.test.ts [hunk 3]
```diff
@@ -437,7 +437,7 @@ apiDescribe('Validation:', persistence => {
         return expectSetToFail(
           db,
           { foo: firebase.firestore.FieldValue.delete() },
-          'FieldValue.delete() can only be used with update() and set() with {merge:true} (found in field foo)'
+          'FieldValue.delete() cannot be used with set() unless you pass {merge:true}'
```
Comment (schmidt-sebastian) [inline packages/firestore/test/integration/api/validation.test.ts]: Why did the field name disappear?

File: packages/firestore/src/api/user_data_converter.ts (hunk 1 above)
Comment (mikelehen) [reply packages/firestore/src/api/user_data_converter.ts]: Good idea, done, though that led to a bit more refactoring to uniformly handle context.path being null (sorry).

File: packages/firestore/src/api/user_data_converter.ts (hunk 2 above)
Comment (mikelehen) [reply packages/firestore/src/api/user_data_converter.ts]: Done.

File: packages/firestore/test/integration/api/validation.test.ts (hunk 3 above)
Comment (mikelehen) [reply packages/firestore/test/integration/api/validation.test.ts]: Uh, turns out it didn't...  But I accidentally removed it from the test and chai's .to.throw() expectation just verifies that the error _contains_ the specified string, so it still passed.

We should probably look into making the expectations more strict, but I think we'd need to switch to using a regex and we'd need to do escaping and stuff, so it's a bit of a pain.  _contains_ is probably good enough for now.  

Fixed the test in any case.

File: packages/firestore/src/api/user_data_converter.ts [hunk 4]
```diff
@@ -529,54 +532,61 @@ export class UserDataConverter {
...
+          'FieldValue.delete() cannot be used with set() unless you pass ' +
+            '{merge:true}'
+        );
+      }
+    } else if (value instanceof ServerTimestampFieldValueImpl) {
+      context.fieldTransforms.push(
+        new FieldTransform(context.path, ServerTimestampTransform.instance)
+      );
+    } else {
+      return fail('Unknown FieldValue type: ' + value);
```
Comment (schmidt-sebastian) [inline packages/firestore/src/api/user_data_converter.ts]: You should be able to remove the `return` since the function now returns `void`.

**Pull request merged**